libbeat/outputs/logstash/deadlock.go

```diff
@@ -0,0 +1,95 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package logstash
+
+import (
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type deadlockListener struct {
+	log     *logp.Logger
+	timeout time.Duration
+	ticker  *time.Ticker
+
+	ackChan chan int
+
+	doneChan chan struct{}
+}
+
+const logstashDeadlockTimeout = 5 * time.Minute
```

Comment:
I wonder if we should make this user configurable with a 5m default.  I think 5m is reasonable, but an admin might come up with a valid reason to increase or decrease.

Reply:
I'd push back on making it configurable in this pass: This error log isn't a solution for the underlying problem, the goal is to drive users towards the real solution (upgrade to a Logstash version that doesn't leave deadlocked connections live, or change the Logstash configuration to use the V1 pipeline bus that doesn't have this bug, or otherwise address whatever issue is leading to crashed Logstash nodes in the first place). Once someone knows they're affected by the bug, there isn't much to be gained by adjusting the exact interval at which the log messages recur, because it wouldn't affect the ingestion behavior. (If it turns out this is a common enough issue that we decide to go ahead with a pipeline workaround, then it would make sense to include a configurable timeout, since that would affect the actual recovery behavior and not just the latency for a particular logged error, but I think for now it would just lock us into a new mostly-inert config flag.)

Instead, I've made the error message more verbose, with suggested solutions and a link to the original issue. Let me know if you think this is a good compromise.

Reply:
I can live with it either way.  

The desire to make it configurable is because we don't know if 5 min **is** a failure case.  Maybe there is a very good reason why a batch, once a day, takes 5m3s.  In that case we log an error message once a day when things are working correctly.  And if the time isn't configurable the admin can't do anything to silence the error message.  That being said I **think** 5 min is large enough that it probably won't be an issue.